File: rust/examples/bench.rs
//! Four timed CPU runs of one text, as one JSON record on stdout.

use loudkit::engine::{Engine, Options};
use serde_json::json;
use std::{env, time::Instant};

fn main() -> Result<(), String> {
    let args: Vec<_> = env::args().collect();
    if args.len() != 4 {
        // A usage line and exit 2, not `assert_eq!`: wrong arguments are a
        // usage error, and a panic answers them with a backtrace hint. The
        // same exit code `src/main.rs` uses for the same case.
        eprintln!("usage: bench BUNDLE VOICE TEXT");
        std::process::exit(2);
    }
    let mut start = Instant::now();
    let mut engine = Engine::load_with(
        &args[1],
        &loudkit::execution::ExecutionConfig {
            onnx_provider: loudkit::execution::OnnxProvider::Cpu,
        },
    )?;
    let load = start.elapsed().as_secs_f64();
    // The checkpoint's rate, not 24000: at any other rate the hardcoded
    // divisor reports the wrong duration for every run.
    let sample_rate = f64::from(u32::try_from(engine.config().sample_rate).unwrap_or(u32::MAX));
    let voice = loudkit::voice::load(&args[2])?;
    let mut runs = Vec::new();
    for run in 0..4 {
        start = Instant::now();
        let (mut first, mut samples, mut tokens, mut chunks) = (0., 0, 0, 0);
        engine.stream(
            &args[3],
            &voice,
            &Options {
                seed: 7,
                ..Default::default()
            },
            None,
            &mut |c| {
                if chunks == 0 {
                    first = start.elapsed().as_secs_f64();
                }
                chunks += 1;
                samples += c.audio.len();
                tokens += c.tokens.len();
                true
            },
        )?;
        runs.push(
            json!({"run":run,"seconds":start.elapsed().as_secs_f64(),"ttfa_s":first,
                        "audio_s":samples as f64/sample_rate,"tokens":tokens,"chunks":chunks}),
        );
    }
    println!(
        "{}",
        json!({"runtime":"rust","bundle":args[1],"load_s":load,"text":args[3],"seed":7,"runs":runs})
    );
    Ok(())
}

File: tools/bench_ports/bench.swift
// Compile with the release LoudKit/LoudKitText objects; see README.md.
import Foundation
import LoudKit
let args = CommandLine.arguments
// A trap here would be the first thing a reader of README.md meets, and this
// file is built with `-O`, where a failed `precondition` prints nothing at all:
// the shell reports "Illegal instruction: 4" and a status of 132. The other
// three runners answer a wrong argument count with the usage line and a status,
// so this one does too.
guard args.count == 4 else {
    FileHandle.standardError.write(Data("usage: bench BUNDLE VOICE TEXT\n".utf8))
    exit(2)
}
let bundle = URL(fileURLWithPath: args[1])
var start = ProcessInfo.processInfo.systemUptime
let engine = try Engine.load(checkpoint: bundle.appendingPathComponent(bundle.lastPathComponent + ".safetensors"))
let load = ProcessInfo.processInfo.systemUptime - start
// The checkpoint's rate, not a literal 24000: at any other rate a hardcoded
// divisor reports the wrong audio duration, and so the wrong RTF, for every
// run. The Rust runner reads it from the engine for the same reason.
let sampleRate = Double(engine.algorithm.sampleRate)
let voice = try VoiceProfile.load(url: URL(fileURLWithPath: args[2]))
var runs: [[String: Any]] = []
for run in 0..<4 {
    start = ProcessInfo.processInfo.systemUptime
    var first = 0.0, samples = 0, tokens = 0, chunks = 0
    try engine.stream(args[3], voice: voice, seed: 7) { chunk in
        if chunks == 0 { first = ProcessInfo.processInfo.systemUptime - start }
        chunks += 1; samples += chunk.audio.count; tokens += chunk.tokens.count
        return true
    }
    runs.append(["run": run, "seconds": ProcessInfo.processInfo.systemUptime - start,
                 "ttfa_s": first, "audio_s": Double(samples) / sampleRate,
                 "tokens": tokens, "chunks": chunks])
}
let output: [String: Any] = ["runtime": "swift", "bundle": args[1], "load_s": load,
                            "text": args[3], "seed": 7, "runs": runs]
print(String(data: try JSONSerialization.data(withJSONObject: output, options: [.sortedKeys]), encoding: .utf8)!)
